### PR TITLE
SQOOP-3396: Add parquet numeric support for Parquet in Hive import

### PR DESCRIPTION
**Pull request by fszabo2**

### CODE REVIEW

File: src/test/org/apache/sqoop/hive/numerictypes/NumericTypesHiveImportTest.java [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+      super(new MysqlImportJobTestConfiguration(), NumericTypesTestUtils.SUCCEED_WITHOUT_EXTRA_ARGS, NumericTypesTestUtils.SUCCEED_WITH_PADDING_ONLY);
+    }
+
+    @Override
+    public DatabaseAdapter createAdapter() {
+      return new MysqlDatabaseAdapter();
+    }
+
+    @Override
+    protected HiveMiniCluster getHiveMiniCluster() {
```
Comment (maugly24) [inline src/test/org/apache/sqoop/hive/numerictypes/NumericTypesHiveImportTest.java]: I wold move getHiveMiniCluster and getHiveServer2TestUtil to the base class, with the concrete implementation, and just would pass the references in the super constructor call. Otherwise you'll have to keep this duplication on all 4 underlying implementation (MySQL, Postgres, MsSQL, Oracle), which sounds quite bad.

File: src/test/org/apache/sqoop/importjob/configuration/MysqlImportJobTestConfiguration.java [hunk 2]
```diff
@@ -65,4 +66,21 @@
   public String toString() {
     return getClass().getSimpleName();
   }
+
+  @Override
+  public Object[] getExpectedResultsForHive() {
```
Comment (maugly24) [inline src/test/org/apache/sqoop/importjob/configuration/MysqlImportJobTestConfiguration.java:82]: Source of code duplication again!
I would move these expected values into a helper class or a super class (choice is up to you).

File: src/test/org/apache/sqoop/importjob/configuration/OracleImportJobTestConfigurationForNumber.java [hunk 3]
```diff
@@ -68,4 +68,14 @@
   public String toString() {
     return getClass().getSimpleName();
   }
+
+  @Override
+  public Object[] getExpectedResultsForHive() {
```
Comment (maugly24) [inline src/test/org/apache/sqoop/importjob/configuration/OracleImportJobTestConfigurationForNumber.java:73]: Same goes here for the duplication. Extract this part please!

File: src/java/org/apache/sqoop/hive/HiveTypes.java [hunk 4]
```diff
@@ -37,16 +42,28 @@
...
+    return toHiveType(sqlType);
+  }
+
+
   /**
    * Given JDBC SQL types coming from another database, what is the best
    * mapping to a Hive-specific type?
    */
-  public static String toHiveType(int sqlType) {
+  private static String toHiveType(int sqlType) {
```
Comment (maugly24) [inline src/java/org/apache/sqoop/hive/HiveTypes.java]: If you touch this class+method anyway would you mind modifying the underlying design in a way not to return null value?

It would be quite error prone e.g. in case of TableDefWriter if the FileLayout is Parquet and by any mistake the type of a given column is not mapped then the execution path would not end up in "not supported" exception, but TableDefWriter would try to create the table with "NULL" datatype.

IMHO this would be a great improvement here.

I would also standardize it in a way not to throw IOException (e.g. getHiveColumnTypeForTextTable) and RuntimeException for the same problem, but go only with RuntimeException.

**Review by maugly24 — CHANGES_REQUESTED** [review]
Hi Fero,

I've found a few issue.

Would you please modify your implementation accordingly?

Thanks,
Attila

File: src/java/org/apache/sqoop/hive/HiveTypes.java [hunk 5]
```diff
@@ -37,16 +42,28 @@
...
+  private static final String HIVE_TYPE_DECIMAL = "DECIMAL";
 
   public static final Log LOG = LogFactory.getLog(HiveTypes.class.getName());
 
   private HiveTypes() { }
 
+
+  public static String toHiveType(int sqlType, SqoopOptions options) {
+
+    if (options.getConf().getBoolean(ConfigurationConstants.PROP_ENABLE_PARQUET_LOGICAL_TYPE_DECIMAL, false)
```
Comment (szvasas) [inline src/java/org/apache/sqoop/hive/HiveTypes.java]: I think this check could go to the original toHiveType to the NUMERIC/DECIMAL switch branch in a separate method.

File: src/test/org/apache/sqoop/hive/numerictypes/NumericTypesHiveImportTest.java (hunk 1 above)
Comment (fszabo2) [reply src/test/org/apache/sqoop/hive/numerictypes/NumericTypesHiveImportTest.java]: Makes sense!

File: src/test/org/apache/sqoop/importjob/configuration/MysqlImportJobTestConfiguration.java (hunk 2 above)
Comment (fszabo2) [reply src/test/org/apache/sqoop/importjob/configuration/MysqlImportJobTestConfiguration.java:82]: Hi Attila, 

I'd like to leave these as is.

I'd argue that these don't actually belong together, so it's not really code duplication but rather an accidental match.

I added a new column to the sample data of the mysql configuration to cover an edge case. The expected result in Hive changed for this configuration, and, on the other hand this edge case does not make sense to be added to the SQL server tests.

File: src/java/org/apache/sqoop/hive/HiveTypes.java (hunk 5 above)
Comment (fszabo2) [reply src/java/org/apache/sqoop/hive/HiveTypes.java]: I refactored this file.

File: src/java/org/apache/sqoop/hive/HiveTypes.java (hunk 4 above)
Comment (fszabo2) [reply src/java/org/apache/sqoop/hive/HiveTypes.java]: Actually, there is a method that defines extra mappings for hive types in OraOop:
org.apache.sqoop.manager.oracle.OraOopConnManager#toHiveType 

So we need to keep the "null" return value here, in order to allow for that to work. 

I think it still makes sense to log a warning though.

On the long run: it might make sense to refactor the ora oop connector, but I think it's out of the scope of this change.

File: src/java/org/apache/sqoop/hive/HiveTypes.java [hunk 6]
```diff
@@ -79,8 +85,42 @@ public static String toHiveType(int sqlType) {
...
         // TODO(aaron): Support BINARY, VARBINARY, LONGVARBINARY, DISTINCT,
         // BLOB, ARRAY, STRUCT, REF, JAVA_OBJECT.
-        return null;
+            return null;
+      }
+  }
+
+  private static String mapDecimalsToHiveType(int sqlType, SqoopOptions options) {
+    if (options.getConf().getBoolean(ConfigurationConstants.PROP_ENABLE_PARQUET_LOGICAL_TYPE_DECIMAL, false)
+        && (sqlType == Types.NUMERIC || sqlType == Types.DECIMAL)){
```
Comment (szvasas) [inline src/java/org/apache/sqoop/hive/HiveTypes.java]: Do we need to test for NUMERIC and DECIMAL types again here?

File: src/java/org/apache/sqoop/hive/HiveTypes.java [hunk 7]
```diff
@@ -79,8 +85,42 @@ public static String toHiveType(int sqlType) {
...
+      return HIVE_TYPE_DECIMAL;
+    }
+    return HIVE_TYPE_DOUBLE;
+  }
+
+
+  public static String toHiveType(Schema schema) {
+    if (schema.getType() == Schema.Type.UNION) {
+      for (Schema subSchema : schema.getTypes()) {
+        if (subSchema.getLogicalType() != null && subSchema.getLogicalType() instanceof Decimal) {
```
Comment (szvasas) [inline src/java/org/apache/sqoop/hive/HiveTypes.java]: Redundant null check: if subSchema.getLogicalType() is null then subSchema.getLogicalType() instanceof Decimal will be false

File: src/test/org/apache/sqoop/importjob/configuration/MysqlImportJobTestConfiguration.java (hunk 2 above)
Comment (maugly24) [reply src/test/org/apache/sqoop/importjob/configuration/MysqlImportJobTestConfiguration.java:82]: Hi @fszabo2 ,

If that's the case:
Could you please at least add some information about the intention behind the values (A.K.A. explanatory comments by Uncle Bob), b/c right now for example if this test would fail I would not have much clue about what values do we have here, and why.

I'm thinking about something like:
this value is an int, XXX precision, so it should be YYYYY in Parquet/Hive

or something like
this value represents the biggest the longest possible values stored as type ZZZ, thus it contains N digits, and should be KKKK in Hive.

Could we add these information, thus put these "Magic numbers" into a context?

Comment (fszabo2) [reply src/test/org/apache/sqoop/importjob/configuration/MysqlImportJobTestConfiguration.java:82]: Fair point!

I added a couple of comments, hope they clarify the magic numbers.

File: src/java/org/apache/sqoop/hive/HiveTypes.java (hunk 4 above)
Comment (fszabo2) [reply src/java/org/apache/sqoop/hive/HiveTypes.java]: After taking another look at this file, it turns out that this method is only called textfile import. Since my only intended use case is parquetfile, I reverted this part of the change.

The implementation will be throwing a RuntimeException (for mapping error) and log warnings in case of a parquet files, as you've suggested.

File: src/test/org/apache/sqoop/hive/TestHiveTypesForAvroTypeMapping.java [hunk 8]
```diff
@@ -38,29 +40,30 @@
 public class TestHiveTypesForAvroTypeMapping {
```
Comment (szvasas) [inline src/test/org/apache/sqoop/hive/TestHiveTypesForAvroTypeMapping.java:48]: Can we add more parameters to this test to cover the changes in org.apache.sqoop.hive.HiveTypes#mapToDecimalOrBinary?

File: src/java/org/apache/sqoop/hive/HiveTypes.java [hunk 9]
```diff
@@ -83,27 +89,58 @@ public static String toHiveType(int sqlType) {
...
+      case FIXED:
+        return HIVE_TYPE_BINARY;
+      default:
+        throw new RuntimeException(String.format("There is no Hive type mapping defined for the Avro type of: %s ", avroType.getName()));
+    }
+  }
+
+  private static String mapToDecimalOrBinary(Schema schema, SqoopOptions options) {
+    boolean logicalTypesEnabled = options.getConf().getBoolean(ConfigurationConstants.PROP_ENABLE_PARQUET_LOGICAL_TYPE_DECIMAL, false);
+    if (logicalTypesEnabled && schema.getLogicalType() != null && schema.getLogicalType() instanceof Decimal) {
```
Comment (szvasas) [inline src/java/org/apache/sqoop/hive/HiveTypes.java]: The null check is redundant here.

File: src/java/org/apache/sqoop/hive/HiveTypes.java [hunk 10]
```diff
@@ -83,27 +89,58 @@ public static String toHiveType(int sqlType) {
...
+      Decimal decimal = (Decimal) schema.getLogicalType();
+
+      // trimming precision and scale to Hive's maximum values.
+      int precision = Math.min(HiveDecimal.MAX_PRECISION, decimal.getPrecision());
+      if (precision < decimal.getPrecision()) {
+        LOG.warn("Warning! Precision in the Hive table definition will be smaller than the actual precision of the column on storage! Hive may not be able to read data from this column.");
```
Comment (szvasas) [inline src/java/org/apache/sqoop/hive/HiveTypes.java:133]: I think we mention this potential problem in the documentation somewhere.

File: src/test/org/apache/sqoop/importjob/numerictypes/NumericTypesAvroImportTestBase.java [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+  public static final Log LOG = LogFactory.getLog(NumericTypesAvroImportTestBase.class.getName());
+
+  public NumericTypesAvroImportTestBase(T configuration, boolean failWithoutExtraArgs, boolean failWithPaddingOnly) {
+    super(configuration, failWithoutExtraArgs, failWithPaddingOnly);
+  }
+
+  @Before
+  public void setUp() {
+    super.setUp();
+    tableDirPath = new Path(getWarehouseDir() + "/" + getTableName());
```
Comment (szvasas) [inline src/test/org/apache/sqoop/importjob/numerictypes/NumericTypesAvroImportTestBase.java]: tableDirPath seems to be initialized in super.setUp, do we need this here?

File: src/test/org/apache/sqoop/importjob/numerictypes/NumericTypesImportTestBase.java [hunk 12]
```diff
@@ -65,240 +46,79 @@
...
-  @After
-  public void tearDown() {
-    try {
-      dropTableIfExists(getTableName());
-    } catch (SQLException e) {
-      LOG.warn("Error trying to drop table on tearDown: " + e);
-    }
-    super.tearDown();
-  }
+  public Path tableDirPath;
```
Comment (szvasas) [inline src/test/org/apache/sqoop/importjob/numerictypes/NumericTypesImportTestBase.java]: This field could be protected.

File: src/test/org/apache/sqoop/importjob/numerictypes/NumericTypesParquetImportTestBase.java [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+  public static final Log LOG = LogFactory.getLog(NumericTypesParquetImportTestBase.class.getName());
+
+  public NumericTypesParquetImportTestBase(T configuration, boolean failWithoutExtraArgs, boolean failWithPaddingOnly) {
+    super(configuration, failWithoutExtraArgs, failWithPaddingOnly);
+  }
+
+  @Before
+  public void setUp() {
+    super.setUp();
+    tableDirPath = new Path(getWarehouseDir() + "/" + getTableName());
```
Comment (szvasas) [inline src/test/org/apache/sqoop/importjob/numerictypes/NumericTypesParquetImportTestBase.java]: tableDirPath seems to be initialized in super.setUp, do we need this here?

File: src/test/org/apache/sqoop/importjob/numerictypes/NumericTypesImportTestBase.java [hunk 14]
```diff
@@ -65,240 +46,79 @@
...
-  }
 
-  private void configureJunitToExpectFailure(boolean failWithPadding) {
-    if (failWithPadding) {
+  @Test
+  public void testImportWithoutPadding() throws IOException {
+    if(failWithoutExtraArgs){
       thrown.expect(IOException.class);
       thrown.expectMessage("Failure during job; return status 1");
+      NumericTypesTestUtils.configureJunitToExpectFailure(thrown);
```
Comment (szvasas) [inline src/test/org/apache/sqoop/importjob/numerictypes/NumericTypesImportTestBase.java:88]: This method seems to be doing the same thing as the previous line.

File: src/test/org/apache/sqoop/importjob/numerictypes/NumericTypesImportTestBase.java [hunk 15]
```diff
@@ -65,240 +46,79 @@
...
-    addEnableParquetDecimal(builder);
-    String[] args = builder.build();
-    runImport(args);
-    if (!failWithoutExtraArgs) {
-      verify(ParquetFile);
+  public void testImportWithPadding() throws IOException {
+    if(failWithPadding){
+      thrown.expect(IOException.class);
+      thrown.expectMessage("Failure during job; return status 1");
+      NumericTypesTestUtils.configureJunitToExpectFailure(thrown);
```
Comment (szvasas) [inline src/test/org/apache/sqoop/importjob/numerictypes/NumericTypesImportTestBase.java:101]: This method seems to be doing the same thing as the previous line.

**Review by szvasas — CHANGES_REQUESTED** [review]

File: src/java/org/apache/sqoop/hive/HiveTypes.java (hunk 6 above)
Comment (fszabo2) [reply src/java/org/apache/sqoop/hive/HiveTypes.java]: This piece of code was reverted.

File: src/java/org/apache/sqoop/hive/HiveTypes.java (hunk 7 above)
Comment (fszabo2) [reply src/java/org/apache/sqoop/hive/HiveTypes.java]: I learn something new every day :), removed.

File: src/java/org/apache/sqoop/hive/HiveTypes.java (hunk 10 above)
Comment (fszabo2) [reply src/java/org/apache/sqoop/hive/HiveTypes.java:133]: Do you think we should remove this warning? (I think, even if it's redundant, it's useful to write this out.)

File: src/test/org/apache/sqoop/importjob/numerictypes/NumericTypesAvroImportTestBase.java (hunk 11 above)
Comment (fszabo2) [reply src/test/org/apache/sqoop/importjob/numerictypes/NumericTypesAvroImportTestBase.java]: Nope.

File: src/test/org/apache/sqoop/importjob/numerictypes/NumericTypesImportTestBase.java (hunk 12 above)
Comment (fszabo2) [reply src/test/org/apache/sqoop/importjob/numerictypes/NumericTypesImportTestBase.java]: Makes sense

File: src/test/org/apache/sqoop/importjob/numerictypes/NumericTypesParquetImportTestBase.java (hunk 13 above)
Comment (fszabo2) [reply src/test/org/apache/sqoop/importjob/numerictypes/NumericTypesParquetImportTestBase.java]: nope.

File: src/test/org/apache/sqoop/importjob/numerictypes/NumericTypesImportTestBase.java (hunk 15 above)
Comment (fszabo2) [reply src/test/org/apache/sqoop/importjob/numerictypes/NumericTypesImportTestBase.java:101]: Indeed!

File: src/test/org/apache/sqoop/importjob/numerictypes/NumericTypesImportTestBase.java (hunk 14 above)
Comment (fszabo2) [reply src/test/org/apache/sqoop/importjob/numerictypes/NumericTypesImportTestBase.java:88]: Indeed! Removed lines above.

File: src/java/org/apache/sqoop/hive/HiveTypes.java (hunk 9 above)
Comment (fszabo2) [reply src/java/org/apache/sqoop/hive/HiveTypes.java]: I'm learning something new every day! :) removed

File: src/java/org/apache/sqoop/hive/HiveTypes.java (hunk 10 above)
Comment (szvasas) [reply src/java/org/apache/sqoop/hive/HiveTypes.java:133]: Sorry, I meant that apart from the warning messages here we should mention it in the documentation too.

File: src/test/org/apache/sqoop/hive/TestHiveTypesForAvroTypeMapping.java [hunk 16]
```diff
@@ -41,30 +44,49 @@
 
   private final String hiveType;
   private final Schema schema;
+  private final SqoopOptions options;
 
-  @Parameters(name = "hiveType = {0}, schema = {1}")
+  @Parameters(name = "hiveType = {0}, schema = {1}, options = {2}")
```
Comment (szvasas) [inline src/test/org/apache/sqoop/hive/TestHiveTypesForAvroTypeMapping.java]: I would not add options here, because SqoopOptions does not have a nice toString() method so we won't see anything meaningful in the test output.

File: src/test/org/apache/sqoop/hive/TestHiveTypesForAvroTypeMapping.java [hunk 17]
```diff
@@ -41,30 +44,49 @@
...
-        new Object[]{"INT", Schema.create(Schema.Type.INT)},
-        new Object[]{"BIGINT", Schema.create(Schema.Type.LONG)},
-        new Object[]{"FLOAT", Schema.create(Schema.Type.FLOAT)},
-        new Object[]{"DOUBLE", Schema.create(Schema.Type.DOUBLE)},
-        new Object[]{"STRING", Schema.createEnum("ENUM", "doc", "namespce", new ArrayList<>())}, // Schema.Type.ENUM
-        new Object[]{"STRING", Schema.create(Schema.Type.STRING)},
-        new Object[]{"BINARY", Schema.create(Schema.Type.BYTES)},
-        new Object[]{"BINARY", Schema.createFixed("Fixed", "doc", "space", 1) }
-        //, new Object[]{"DECIMAL", Schema.create(Schema.Type.UNION).}
+        new Object[]{"BOOLEAN", Schema.create(Schema.Type.BOOLEAN), new SqoopOptions()},
```
Comment (szvasas) [inline src/test/org/apache/sqoop/hive/TestHiveTypesForAvroTypeMapping.java]: We could static import the create methods from Schema so that would make these lines shorter.

File: src/test/org/apache/sqoop/hive/TestHiveTypesForAvroTypeMapping.java [hunk 18]
```diff
@@ -41,30 +44,49 @@
...
+        new Object[]{"INT", Schema.create(Schema.Type.INT), new SqoopOptions()},
+        new Object[]{"BIGINT", Schema.create(Schema.Type.LONG), new SqoopOptions()},
+        new Object[]{"FLOAT", Schema.create(Schema.Type.FLOAT), new SqoopOptions()},
+        new Object[]{"DOUBLE", Schema.create(Schema.Type.DOUBLE), new SqoopOptions()},
+        new Object[]{"STRING", Schema.createEnum("ENUM", "doc", "namespce", new ArrayList<>()), new SqoopOptions()}, // Schema.Type.ENUM
```
Comment (szvasas) [inline src/test/org/apache/sqoop/hive/TestHiveTypesForAvroTypeMapping.java]: Do we need the comment at the end of the line?

File: src/test/org/apache/sqoop/hive/TestHiveTypesForAvroTypeMapping.java [hunk 19]
```diff
@@ -41,30 +44,49 @@
...
+        new Object[]{"STRING", Schema.create(Schema.Type.STRING), new SqoopOptions()},
+        new Object[]{"BINARY", Schema.create(Schema.Type.BYTES), new SqoopOptions()},
```
Comment (szvasas) [inline src/test/org/apache/sqoop/hive/TestHiveTypesForAvroTypeMapping.java]: I would add one more test case here:
new Object[]{"BINARY", create(Schema.Type.BYTES), createSqoopOptionsWithLogicalTypesEnabled()},

to make sure that we test the if condition org/apache/sqoop/hive/HiveTypes.java:127 fully

**Review by szvasas — COMMENTED** [review]
Thanks for improving this test case, I have just few minor comments.

File: src/java/org/apache/sqoop/hive/HiveTypes.java (hunk 10 above)
Comment (fszabo2) [reply src/java/org/apache/sqoop/hive/HiveTypes.java:133]: I created SQOOP-3418 to cover the documentation part.

File: src/test/org/apache/sqoop/hive/TestHiveTypesForAvroTypeMapping.java (hunk 16 above)
Comment (fszabo2) [reply src/test/org/apache/sqoop/hive/TestHiveTypesForAvroTypeMapping.java]: removed

File: src/test/org/apache/sqoop/hive/TestHiveTypesForAvroTypeMapping.java (hunk 18 above)
Comment (fszabo2) [reply src/test/org/apache/sqoop/hive/TestHiveTypesForAvroTypeMapping.java]: removed

File: src/test/org/apache/sqoop/hive/TestHiveTypesForAvroTypeMapping.java (hunk 17 above)
Comment (fszabo2) [reply src/test/org/apache/sqoop/hive/TestHiveTypesForAvroTypeMapping.java]: I've added a bunch of static imports (also for Schema.Type) so now the whole thing is a lot more readable.

File: src/test/org/apache/sqoop/hive/TestHiveTypesForAvroTypeMapping.java (hunk 19 above)
Comment (fszabo2) [reply src/test/org/apache/sqoop/hive/TestHiveTypesForAvroTypeMapping.java]: Yeah, makes sense.

File: src/test/org/apache/sqoop/importjob/numerictypes/NumericTypesParquetImportTestBase.java [hunk 20]
```diff
@@ -0,0 +1,83 @@
...
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.hadoop.fs.Path;
+import org.apache.parquet.schema.MessageType;
+import org.apache.parquet.schema.OriginalType;
+import org.apache.sqoop.importjob.configuration.ParquetTestConfiguration;
+import org.apache.sqoop.testutil.ArgumentArrayBuilder;
+import org.apache.sqoop.testutil.NumericTypesTestUtils;
+import org.apache.sqoop.util.ParquetReader;
+import org.junit.Before;
```
Comment (szvasas) [inline src/test/org/apache/sqoop/importjob/numerictypes/NumericTypesParquetImportTestBase.java]: Unused import

File: src/test/org/apache/sqoop/importjob/numerictypes/NumericTypesParquetImportTestBase.java [hunk 21]
```diff
@@ -0,0 +1,83 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.sqoop.importjob.numerictypes;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.hadoop.fs.Path;
```
Comment (szvasas) [inline src/test/org/apache/sqoop/importjob/numerictypes/NumericTypesParquetImportTestBase.java]: Unused import

**Review by szvasas — APPROVED** [review]
